torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -5119,3 +5119,88 @@ def test_output_unused_in_loss(self):
                                 )
                         else:
                             self.assertFalse(True, "DDP error not raised")
+
+        @require_backend({"gloo"})
+        @require_backends_available({"gloo"})
+        def test_monitored_barrier_gloo(self):
+            process_group = dist.new_group(ranks=list(i for i in range(int(self.world_size))))
+            tensors = [torch.ones(10) * self.rank]
+            # Kick off some allreduce work on all ranks
+            for _ in range(10):
+                process_group.allreduce(tensors).wait()
+            # Run monitored barrier
+            timeout = timedelta(seconds=0.1)
+            process_group.monitored_barrier(timeout)
+            # All ranks besides 1 call into barrier, rank 0 should report failure
+            # while others report gloo error.
+            failed_rank = 1
+            if self.rank == failed_rank:
+                return
+            if self.rank == 0:
+                with self.assertRaisesRegex(RuntimeError, f"Rank {failed_rank}"):
+                    process_group.monitored_barrier(timeout)
+            else:
+                # It is permissible for other ranks that did not fail to
+                # successfully exit or crash in the monitored barrier, the main
+                # purpose of monitored barrier is to report the rank that hung
+                # on rank 0.
+                try:
+                    process_group.monitored_barrier(timeout)
+                except RuntimeError:
+                    pass
+
+        @require_backend({"gloo", "nccl"})
+        @require_backends_available({"gloo", "nccl"})
+        def test_monitored_barrier_allreduce_hang(self):
+            # tests expected behavior when nonzero rank hangs.
+            if "NCCL_ASYNC_ERROR_HANDLING" in os.environ:
+                del os.environ["NCCL_ASYNC_ERROR_HANDLING"]
+            os.environ["NCCL_BLOCKING_WAIT"] = "1"
+            nccl_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                timeout=timedelta(seconds=2),
```

Comment:
We use a smaller timeout in the allreduce that we expect to fail, however this one should be generous enough for the allreduce before that which we want to succeed.